File: backend/app/schemas/dxcluster.py
"""DX Cluster 相关 Pydantic 模型。"""

from pydantic import BaseModel, Field
from typing import Optional, Any


class SpotResponse(BaseModel):
    """单条 Spot 响应。"""
    spotter: str = Field(..., description="通报者呼号（发 spot 的人）")
    freq: float = Field(..., description="频率 (MHz)")
    dx_callsign: str = Field(..., description="被叫台呼号（通联目标）")
    mode: str = Field("", description="模式（FT8/CW/SSB 等）")
    comment: str = Field("", description="spotter 注释")
    time_utc: Optional[str] = Field(None, description="原始 UTC 时间字符串")
    band: str = Field("", description="由频率推算的业余波段")
    received_at: str = Field(..., description="本服务收到该 spot 的 UTC ISO 时间")


class NodeInfoResponse(BaseModel):
    """DX Cluster 节点信息。"""
    host: str
    port: int
    name: str
    country: str
    remark: str = ""


class ClusterStatusResponse(BaseModel):
    """DX Cluster 连接状态。"""
    connected: bool
    connecting: bool = False
    current_node: Optional[NodeInfoResponse] = None
    callsign: Optional[str] = None
    spot_count: int = 0
    uptime_seconds: Optional[float] = None


class SwitchNodeRequest(BaseModel):
    """连接/切换节点请求。"""
    node_host: str = Field(..., description="节点主机地址")
    node_port: int = Field(..., ge=1, le=65535, description="节点端口")


class ConnectResult(BaseModel):
    """连接操作结果。"""
    success: bool
    message: str = ""
    status: Optional[ClusterStatusResponse] = None

File: frontend/src/api/dxcluster.ts
import api from './index'

/** DX Cluster 节点信息 */
export interface DxNode {
  host: string
  port: number
  name: string
  country: string
  remark: string
}

/** 单条 Spot */
export interface Spot {
  spotter: string
  freq: number
  dx_callsign: string
  mode: string
  comment: string
  time_utc: string | null
  band: string
  received_at: string
}

/** 连接状态 */
export interface ClusterStatus {
  connected: boolean
  connecting: boolean
  current_node: DxNode | null
  callsign: string | null
  spot_count: number
  uptime_seconds: number | null
}

export const dxclusterApi = {
  getNodes(): Promise<DxNode[]> {
    return api.get('/dxcluster/nodes')
  },

  getStatus(): Promise<ClusterStatus> {
    return api.get('/dxcluster/status')
  },

  getSpots(limit = 50): Promise<Spot[]> {
    return api.get('/dxcluster/spots', { params: { limit } })
  },

  connect(nodeHost: string, nodePort: number): Promise<{ success: boolean; message: string; status: ClusterStatus }> {
    return api.post('/dxcluster/connect', { node_host: nodeHost, node_port: nodePort })
  },

  disconnect(): Promise<ClusterStatus> {
    return api.post('/dxcluster/disconnect')
  },
}
